index.html

```diff
@@ -2543,18 +2588,271 @@ <h1>
+                    </dl>
+                    
+                </dd>
+            
+            </dl>
+        </p>
+        
+        <p>
+The <code>content</code> of this function's output is a byte stream of data. This byte stream MAY 
+represent a <a>DID document</a> in a conformant representation, a portion of a <a>DID document</a>
+in a conformant representation, the results of calling a service
+endpoint, or something else determined by the <a>DID method</a> and its services.
+        </p>
+        
+        <p>
+If the returned <code>content</code> is a <a>DID document</a>, the <a>DID URL dereferencer</a> MAY
+return an additional set of DID document metadata, as described in <a href="#did-resolution"></a>.
+        </p>
+        
+    </section>
+
+    <section>
+        <h2>
+Metadata structure
+        </h2>
+
+        <div class="issue" data-number="58">
+There needs to be a registry established for handling input options and rules defined for 
+how to handle an unknown input option. A similar issue applies for all metadata structures
+returned from these functions. 
+        </div>
+        
+        <p>
+Metadata structures are maps of key-value string pairs. All keys within this map
+MUST be UTF-8 unicode strings. The keys of this map MUST NOT be repeated within 
+a given metadata structure. The keys of this map MUST be compared in a 
+case-insensitive way and the case MAY be transformed. The values of this map 
+MUST be retained as an exact string without transformation. 
+        </p>
+
+        <p>
+All <a>bindings</a> MUST provide a way for the caller of the function to input or 
+receive the map of key-value string pairs as described in this specification. The 
+exact representation of these metadata structures as part of a protocol or serialization
+is out of scope for this specification. For example, an HTTP-based binding can
+use HTTP headers of an HTTP request to convey input options and HTTP headers of an
+HTTP response to convey document and resolver metadata. Alternatively, a fully
+JSON-based binding can use a JSON object to represent the metadata structures.
+        </p>
+        
+        <p>
+The keys and values MAY be transformed by the <a>binding</a> for transport 
+in a deterministic and reversible fashion. For example, the <code>content-type</code>
+<a>DID resolution</a> metadata element could be represented as a text header
+in a binding:
+        </p>
+        
+        <p>
+<code>DID-CONTENT-TYPE: application/json+did</code>
+        </p>
+        
+        <p>
+In this example, the key <code>content-type</code> has had the string <code>DID-</code>
+prepended to it and the key text is uppercased. When returned from the function, this transformation
+would be reversed to provide the value <code>application/json+did</code> under the key
+<code>content-type</code> to the client. In other words, the client MUST NOT need
```

Comment:
"client" and "caller" are being used interchangeably in this PR and it's confusing if something different is meant when one is used over the other. If there is a difference, it's so nuanced that I can't tell. Are we talking about clients, callers, or both?

Reply:
How would we test this "MUST NOT" normative statement? I'm struggling to think about what test would be written.

Reply:
The client is defined as the component that calls the function, above. However it is awkward to say the "client of the function". Would appreciate suggested text.

Reply:
If your argument is that we're merely outlining functions here and not protocol - then I suggest staying away from client. Alternatives include "caller of the function", "when the function is executed" or "consumer of the output".

Out of all of those, I'd prefer "when the function is executed", but it doesn't work here. Something like the following might be better:

> In other words, a consumer of <a>DID resolution</a> output MUST NOT need to know the underlying representation of the metadata used by the implementation of the function.

Doing so doesn't give me any more insight into how we'd write a test in a test suite for this item.

Reply:
I'm fine with having more precise definitions to use here, as **client** is a fairly generic term. 

I'm not sure what you mean about the test suite. These function definitions actually make it much easier to write a test suite, and in fact make it easier to test the entire rest of the specification. You simply have the test suite implement a hook for `resolve` which is then implemented by whatever adapters you need to call different methods to do the actual resolution. The test suite code then doesn't need to change whether it's testing an implementation of the input of a function or the implementation of an output of a function or the implementation of tying those two together. So your test suite would test this by sending in a set of standard or common inputs and expecting a set of standard or common outputs. If those work, it passes. The test suite doesn't care what happens underneath as long as the contract is fulfilled, which is the very nature of a test suite.

Reply:
I've removed the term "client" from this section, even though it's used elsewhere in the document.